File: plugins/chromecast/ios/CastModule.swift
import Foundation
import GoogleCast

/**
 * iOS native module for Google Cast (Chromecast) support.
 * Mirrors the Android CastModule API for cross-platform compatibility.
 */
@objc(CastModule)
class CastModule: RCTEventEmitter {

  private var sessionManager: GCKSessionManager {
    GCKCastContext.sharedInstance().sessionManager
  }

  private var hasListeners = false
  private var statusTimer: Timer?

  override init() {
    super.init()
    setupCastContext()
  }

  private func setupCastContext() {
    let options = GCKCastOptions(discoveryCriteria:
      GCKDiscoveryCriteria(applicationID: kGCKDefaultMediaReceiverApplicationID))
    GCKCastContext.setSharedInstanceWith(options)
    sessionManager.add(self)
  }

  @objc override static func requiresMainQueueSetup() -> Bool { true }

  override func supportedEvents() -> [String] {
    return [
      "onSessionStarted",
      "onSessionEnded",
      "onSessionStartFailed",
      "onMediaStatusUpdate",
      "onDevicesDiscovered",
    ]
  }

  override func startObserving() { hasListeners = true }
  override func stopObserving() {
    hasListeners = false
    stopMediaStatusPolling()
  }

  // MARK: - Discovery

  @objc func startDiscovery(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      GCKCastContext.sharedInstance().discoveryManager.startDiscovery()
      resolve(true)
    }
  }

  @objc func getAvailableDevices(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      let discoveryManager = GCKCastContext.sharedInstance().discoveryManager
      var devices: [[String: Any]] = []
      for i in 0..<discoveryManager.deviceCount {
        let device = discoveryManager.device(at: i)
        devices.append([
          "id": device.deviceID,
          "name": device.friendlyName ?? "Unknown",
          "isConnected": self.sessionManager.currentCastSession?.connectionState == .connected
            && self.sessionManager.currentCastSession?.device?.deviceID == device.deviceID,
        ])
      }
      resolve(devices)
    }
  }

  @objc func showCastPicker(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      GCKCastContext.sharedInstance().presentCastDialog()
      resolve(true)
    }
  }

  // MARK: - Media Control

  @objc func loadMedia(_ url: String, title: String, author: String, coverUrl: String, position: Double,
                       resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        reject("NO_CLIENT", "No remote media client", nil)
        return
      }

      let metadata = GCKMediaMetadata(metadataType: .musicTrack)
      metadata.setString(title, forKey: kGCKMetadataKeyTitle)
      metadata.setString(author, forKey: kGCKMetadataKeyArtist)
      if !coverUrl.isEmpty, let imageUrl = URL(string: coverUrl) {
        metadata.addImage(GCKImage(url: imageUrl, width: 800, height: 800))
      }

      guard let mediaURL = URL(string: url) else {
        reject("INVALID_URL", "Invalid media URL", nil)
        return
      }
      let builder = GCKMediaInformationBuilder(contentURL: mediaURL)
      builder.streamType = .buffered
      builder.contentType = "audio/mp4"
      builder.metadata = metadata

      let options = GCKMediaLoadOptions()
      options.autoplay = true
      options.playPosition = position

      let request = client.loadMedia(builder.build(), with: options)
      request.delegate = RequestDelegate(resolve: resolve, reject: reject)
      self.startMediaStatusPolling()
    }
  }

  @objc func play(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        reject("NO_CLIENT", "No remote media client", nil)
        return
      }
      client.play()
      resolve(true)
    }
  }

  @objc func pause(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        reject("NO_CLIENT", "No remote media client", nil)
        return
      }
      client.pause()
      resolve(true)
    }
  }

  @objc func seek(_ position: Double, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        reject("NO_CLIENT", "No remote media client", nil)
        return
      }
      let options = GCKMediaSeekOptions()
      options.interval = position
      client.seek(with: options)
      resolve(true)
    }
  }

  @objc func stop(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        reject("NO_CLIENT", "No remote media client", nil)
        return
      }
      client.stop()
      resolve(true)
    }
  }

  @objc func disconnect(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      self.sessionManager.endSessionAndStopCasting(true)
      resolve(true)
    }
  }

  @objc func getPosition(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let client = self.sessionManager.currentCastSession?.remoteMediaClient else {
        resolve(-1.0)
        return
      }
      resolve(client.approximateStreamPosition())
    }
  }

  @objc func isConnected(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      resolve(self.sessionManager.currentCastSession?.connectionState == .connected)
    }
  }

  // MARK: - Status Polling

  private func startMediaStatusPolling() {
    stopMediaStatusPolling()
    statusTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
      self?.emitMediaStatus()
    }
  }

  private func stopMediaStatusPolling() {
    statusTimer?.invalidate()
    statusTimer = nil
  }

  private func emitMediaStatus() {
    guard hasListeners,
          let client = sessionManager.currentCastSession?.remoteMediaClient,
          let status = client.mediaStatus else { return }

    sendEvent(withName: "onMediaStatusUpdate", body: [
      "position": client.approximateStreamPosition(),
      "duration": status.mediaInformation?.streamDuration ?? 0,
      "isPlaying": status.playerState == .playing,
      "isPaused": status.playerState == .paused,
      "isIdle": status.playerState == .idle,
    ])
  }
}

// MARK: - GCKSessionManagerListener

extension CastModule: GCKSessionManagerListener {
  func sessionManager(_ sessionManager: GCKSessionManager, didStart session: GCKCastSession) {
    guard hasListeners else { return }
    sendEvent(withName: "onSessionStarted", body: [
      "sessionId": session.sessionID ?? "",
      "deviceName": session.device?.friendlyName ?? "Unknown",
    ])
  }

  func sessionManager(_ sessionManager: GCKSessionManager, didEnd session: GCKCastSession, withError error: Error?) {
    stopMediaStatusPolling()
    guard hasListeners else { return }
    sendEvent(withName: "onSessionEnded", body: [
      "error": error?.localizedDescription ?? "",
    ])
  }

  func sessionManager(_ sessionManager: GCKSessionManager, didFailToStart session: GCKCastSession, withError error: Error) {
    guard hasListeners else { return }
    sendEvent(withName: "onSessionStartFailed", body: [
      "error": error.localizedDescription,
    ])
  }
}

// MARK: - Request Delegate

private class RequestDelegate: NSObject, GCKRequestDelegate {
  let resolve: RCTPromiseResolveBlock
  let reject: RCTPromiseRejectBlock

  init(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    self.resolve = resolve
    self.reject = reject
  }

  func requestDidComplete(_ request: GCKRequest) {
    resolve(true)
  }

  func request(_ request: GCKRequest, didFailWithError error: GCKError) {
    reject("LOAD_ERROR", error.localizedDescription, error)
  }
}

File: plugins/chromecast/src/CastModule.kt
package com.secretlibrary.app.chromecast

import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.google.android.gms.cast.MediaInfo
import com.google.android.gms.cast.MediaLoadRequestData
import com.google.android.gms.cast.MediaMetadata
import com.google.android.gms.cast.MediaStatus
import com.google.android.gms.cast.framework.*
import com.google.android.gms.cast.framework.media.RemoteMediaClient
import com.google.android.gms.common.images.WebImage
import android.net.Uri
import java.util.concurrent.Executors

/**
 * React Native bridge for Google Cast SDK.
 *
 * Uses async CastContext initialization (required for Cast SDK 22+).
 * Shows the native MediaRouteButton picker dialog on tap.
 */
class CastModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "CastModule"
        private const val MODULE_NAME = "CastModule"
    }

    private var castContext: CastContext? = null
    private var sessionManager: SessionManager? = null
    private var mediaClientCallback: RemoteMediaClient.Callback? = null

    private val sessionListener = object : SessionManagerListener<CastSession> {
        override fun onSessionStarting(session: CastSession) {}

        override fun onSessionStarted(session: CastSession, sessionId: String) {
            Log.d(TAG, "Cast session started: $sessionId")
            registerMediaClientCallback(session)
            sendEvent("onSessionStarted", Arguments.createMap().apply {
                putString("sessionId", sessionId)
                putString("deviceName", session.castDevice?.friendlyName ?: "Unknown")
            })
            startMediaStatusPolling()
        }

        override fun onSessionStartFailed(session: CastSession, error: Int) {
            Log.e(TAG, "Cast session start failed: $error")
            sendEvent("onSessionStartFailed", Arguments.createMap().apply {
                putInt("error", error)
            })
        }

        override fun onSessionEnding(session: CastSession) {}

        override fun onSessionEnded(session: CastSession, error: Int) {
            Log.d(TAG, "Cast session ended")
            unregisterMediaClientCallback(session)
            stopMediaStatusPolling()
            sendEvent("onSessionEnded", Arguments.createMap().apply {
                putInt("error", error)
            })
        }

        override fun onSessionResuming(session: CastSession, sessionId: String) {}

        override fun onSessionResumed(session: CastSession, wasSuspended: Boolean) {
            Log.d(TAG, "Cast session resumed")
            registerMediaClientCallback(session)
            sendEvent("onSessionStarted", Arguments.createMap().apply {
                putString("sessionId", session.sessionId ?: "")
                putString("deviceName", session.castDevice?.friendlyName ?: "Unknown")
            })
            startMediaStatusPolling()
        }

        override fun onSessionResumeFailed(session: CastSession, error: Int) {}

        override fun onSessionSuspended(session: CastSession, reason: Int) {}
    }

    override fun getName(): String = MODULE_NAME

    override fun initialize() {
        super.initialize()
        // Use async initialization (Cast SDK 22+ requirement)
        try {
            CastContext.getSharedInstance(reactApplicationContext, Executors.newSingleThreadExecutor())
                .addOnSuccessListener { ctx ->
                    Log.d(TAG, "CastContext initialized successfully")
                    castContext = ctx
                    sessionManager = ctx.sessionManager
                    sessionManager?.addSessionManagerListener(sessionListener, CastSession::class.java)

                    // Check if there's an existing session (app restart while casting)
                    val existingSession = ctx.sessionManager.currentCastSession
                    if (existingSession?.isConnected == true) {
                        Log.d(TAG, "Found existing Cast session on init")
                        registerMediaClientCallback(existingSession)
                        sendEvent("onSessionStarted", Arguments.createMap().apply {
                            putString("sessionId", existingSession.sessionId ?: "")
                            putString("deviceName", existingSession.castDevice?.friendlyName ?: "Unknown")
                        })
                        startMediaStatusPolling()
                    }
                }
                .addOnFailureListener { e ->
                    Log.e(TAG, "CastContext initialization failed", e)
                }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start Cast context initialization", e)
        }
    }

    /**
     * Register a RemoteMediaClient.Callback for immediate state change notifications.
     * This supplements the polling loop with instant events for play/pause/finish/error.
     */
    private fun registerMediaClientCallback(session: CastSession) {
        unregisterMediaClientCallback(session) // prevent duplicates
        val client = session.remoteMediaClient ?: return

        val callback = object : RemoteMediaClient.Callback() {
            override fun onStatusUpdated() {
                val status = client.mediaStatus ?: return
                val playerState = status.playerState

                // Detect playback finished (idle with FINISHED reason)
                if (playerState == MediaStatus.PLAYER_STATE_IDLE) {
                    val idleReason = status.idleReason
                    if (idleReason == MediaStatus.IDLE_REASON_FINISHED) {
                        Log.d(TAG, "Cast media playback finished")
                        sendEvent("onMediaFinished", Arguments.createMap().apply {
                            putDouble("position", client.approximateStreamPosition / 1000.0)
                            putDouble("duration", (status.mediaInfo?.streamDuration ?: 0) / 1000.0)
                        })
                    } else if (idleReason == MediaStatus.IDLE_REASON_ERROR) {
                        Log.e(TAG, "Cast media playback error")
                        sendEvent("onMediaError", Arguments.createMap().apply {
                            putString("error", "Media playback error on Cast device")
                        })
                    }
                }
            }
        }
        mediaClientCallback = callback
        client.registerCallback(callback)
    }

    private fun unregisterMediaClientCallback(session: CastSession) {
        val callback = mediaClientCallback ?: return
        session.remoteMediaClient?.unregisterCallback(callback)
        mediaClientCallback = null
    }

    @Suppress("DEPRECATION")
    override fun onCatalystInstanceDestroy() {
        sessionManager?.currentCastSession?.let { unregisterMediaClientCallback(it) }
        sessionManager?.removeSessionManagerListener(sessionListener, CastSession::class.java)
        stopMediaStatusPolling()
        super.onCatalystInstanceDestroy()
    }

    @ReactMethod
    fun startDiscovery(promise: Promise) {
        try {
            val selector = castContext?.mergedSelector
            promise.resolve(selector != null)
        } catch (e: Exception) {
            promise.reject("DISCOVERY_ERROR", "Failed to start discovery", e)
        }
    }

    @ReactMethod
    fun getAvailableDevices(promise: Promise) {
        try {
            val devices = Arguments.createArray()
            val currentSession = sessionManager?.currentCastSession
            if (currentSession != null) {
                val device = Arguments.createMap().apply {
                    putString("id", currentSession.castDevice?.deviceId ?: "")
                    putString("name", currentSession.castDevice?.friendlyName ?: "Unknown")
                    putBoolean("isConnected", true)
                }
                devices.pushMap(device)
            }
            promise.resolve(devices)
        } catch (e: Exception) {
            promise.reject("DEVICES_ERROR", "Failed to get devices", e)
        }
    }

    /**
     * Show the native Cast device picker dialog.
     *
     * Uses MediaRouteChooserDialog directly (when not connected) or
     * MediaRouteControllerDialog (when connected). This is more reliable
     * than the MediaRouteButton.showDialog() workaround which can fail
     * silently if the button isn't fully attached to the window.
     */
    @ReactMethod
    fun showCastPicker(promise: Promise) {
        val activity = reactApplicationContext.currentActivity
        if (activity == null) {
            Log.e(TAG, "showCastPicker: No activity available")
            promise.reject("NO_ACTIVITY", "No activity available")
            return
        }

        // If async init hasn't completed yet, try synchronous fallback
        var ctx = castContext
        if (ctx == null) {
            Log.w(TAG, "showCastPicker: CastContext not ready, trying synchronous init")
            try {
                @Suppress("DEPRECATION")
                ctx = CastContext.getSharedInstance(activity)
                castContext = ctx
                // Only register listener if initialize() hasn't done it yet
                if (sessionManager == null) {
                    sessionManager = ctx.sessionManager
                    sessionManager?.addSessionManagerListener(sessionListener, CastSession::class.java)
                }
                Log.d(TAG, "showCastPicker: Synchronous CastContext init succeeded")
            } catch (e: Exception) {
                Log.e(TAG, "showCastPicker: CastContext not available", e)
                promise.reject("NO_CAST_CONTEXT", "Cast not initialized yet", e)
                return
            }
        }

        val selector = ctx!!.mergedSelector
        if (selector == null) {
            Log.w(TAG, "showCastPicker: No route selector — Cast SDK may not be configured")
            promise.reject("NO_SELECTOR", "No route selector available")
            return
        }

        activity.runOnUiThread {
            try {
                val session = sessionManager?.currentCastSession
                if (session?.isConnected == true) {
                    // Already connected — show controller dialog
                    Log.d(TAG, "showCastPicker: Showing controller dialog (connected to ${session.castDevice?.friendlyName})")
                    val dialog = androidx.mediarouter.app.MediaRouteControllerDialog(activity)
                    dialog.show()
                } else {
                    // Not connected — show device chooser
                    Log.d(TAG, "showCastPicker: Showing chooser dialog")
                    val dialog = androidx.mediarouter.app.MediaRouteChooserDialog(activity)
                    dialog.routeSelector = selector
                    dialog.show()
                }
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e(TAG, "showCastPicker failed", e)
                promise.reject("PICKER_ERROR", "Failed to show cast picker", e)
            }
        }
    }

    @ReactMethod
    fun loadMedia(url: String, title: String, author: String, coverUrl: String, position: Double, contentType: String, promise: Promise) {
        try {
            val session = sessionManager?.currentCastSession
            if (session == null) {
                promise.reject("NO_SESSION", "No active Cast session")
                return
            }

            val remoteClient = session.remoteMediaClient
            if (remoteClient == null) {
                promise.reject("NO_CLIENT", "No remote media client")
                return
            }

            // Use AUDIOBOOK_CHAPTER type for proper Cast device UI
            val metadata = MediaMetadata(MediaMetadata.MEDIA_TYPE_AUDIOBOOK_CHAPTER).apply {
                putString(MediaMetadata.KEY_TITLE, title)
                putString(MediaMetadata.KEY_ARTIST, author)
                // Also set book title for audiobook displays
                putString(MediaMetadata.KEY_BOOK_TITLE, title)
                if (coverUrl.isNotEmpty()) {
                    addImage(WebImage(Uri.parse(coverUrl)))
                }
            }

            // Use provided content type, fall back to audio/mp4 for m4b files
            val mimeType = if (contentType.isNotEmpty()) contentType else "audio/mp4"

            val mediaInfo = MediaInfo.Builder(url)
                .setStreamType(MediaInfo.STREAM_TYPE_BUFFERED)
                .setContentType(mimeType)
                .setMetadata(metadata)
                .build()

            val loadRequest = MediaLoadRequestData.Builder()
                .setMediaInfo(mediaInfo)
                .setAutoplay(true)
                .setCurrentTime((position * 1000).toLong())
                .build()

            Log.d(TAG, "Loading media: url=${url.take(80)}..., contentType=$mimeType, position=${position}s")

            remoteClient.load(loadRequest)
                .setResultCallback { result ->
                    if (result.status.isSuccess) {
                        Log.d(TAG, "Media loaded successfully on Cast device")
                        promise.resolve(true)
                        startMediaStatusPolling()
                    } else {
                        Log.e(TAG, "Failed to load media: ${result.status.statusMessage}")
                        promise.reject("LOAD_ERROR", "Failed to load media: ${result.status.statusMessage}")
                    }
                }
        } catch (e: Exception) {
            promise.reject("LOAD_ERROR", "Failed to load media", e)
        }
    }

    @ReactMethod
    fun play(promise: Promise) {
        try {
            val client = sessionManager?.currentCastSession?.remoteMediaClient
            if (client == null) {
                promise.reject("NO_CLIENT", "No remote media client")
                return
            }
            client.play()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("PLAY_ERROR", "Failed to play", e)
        }
    }

    @ReactMethod
    fun pause(promise: Promise) {
        try {
            val client = sessionManager?.currentCastSession?.remoteMediaClient
            if (client == null) {
                promise.reject("NO_CLIENT", "No remote media client")
                return
            }
            client.pause()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("PAUSE_ERROR", "Failed to pause", e)
        }
    }

    @ReactMethod
    fun seek(position: Double, promise: Promise) {
        try {
            val client = sessionManager?.currentCastSession?.remoteMediaClient
            if (client == null) {
                promise.reject("NO_CLIENT", "No remote media client")
                return
            }
            client.seek(com.google.android.gms.cast.MediaSeekOptions.Builder()
                .setPosition((position * 1000).toLong())
                .build())
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("SEEK_ERROR", "Failed to seek", e)
        }
    }

    @ReactMethod
    fun stop(promise: Promise) {
        try {
            val client = sessionManager?.currentCastSession?.remoteMediaClient
            if (client == null) {
                promise.reject("NO_CLIENT", "No remote media client")
                return
            }
            client.stop()
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("STOP_ERROR", "Failed to stop", e)
        }
    }

    @ReactMethod
    fun disconnect(promise: Promise) {
        try {
            sessionManager?.endCurrentSession(true)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("DISCONNECT_ERROR", "Failed to disconnect", e)
        }
    }

    @ReactMethod
    fun getPosition(promise: Promise) {
        try {
            val client = sessionManager?.currentCastSession?.remoteMediaClient
            if (client == null) {
                promise.resolve(-1.0)
                return
            }
            promise.resolve(client.approximateStreamPosition / 1000.0)
        } catch (e: Exception) {
            promise.resolve(-1.0)
        }
    }

    @ReactMethod
    fun isConnected(promise: Promise) {
        try {
            val session = sessionManager?.currentCastSession
            promise.resolve(session?.isConnected == true)
        } catch (e: Exception) {
            promise.resolve(false)
        }
    }

    // Required for NativeEventEmitter
    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}

    private var mediaStatusRunnable: Runnable? = null
    private val handler = android.os.Handler(android.os.Looper.getMainLooper())

    private fun startMediaStatusPolling() {
        stopMediaStatusPolling()
        mediaStatusRunnable = object : Runnable {
            override fun run() {
                val client = sessionManager?.currentCastSession?.remoteMediaClient
                if (client != null) {
                    val status = client.mediaStatus
                    if (status != null) {
                        sendEvent("onMediaStatusUpdate", Arguments.createMap().apply {
                            putDouble("position", client.approximateStreamPosition / 1000.0)
                            putDouble("duration", (status.mediaInfo?.streamDuration ?: 0) / 1000.0)
                            putBoolean("isPlaying", status.playerState == MediaStatus.PLAYER_STATE_PLAYING)
                            putBoolean("isPaused", status.playerState == MediaStatus.PLAYER_STATE_PAUSED)
                            putBoolean("isIdle", status.playerState == MediaStatus.PLAYER_STATE_IDLE)
                            putInt("idleReason", status.idleReason)
                        })
                    }
                    handler.postDelayed(this, 1000)
                }
            }
        }
        handler.postDelayed(mediaStatusRunnable!!, 1000)
    }

    private fun stopMediaStatusPolling() {
        mediaStatusRunnable?.let { handler.removeCallbacks(it) }
        mediaStatusRunnable = null
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        try {
            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send event $eventName", e)
        }
    }
}
